### ISSUE TITLE
“Asymmetric transaction rollback error” if commit callbacks throw exceptions

### ISSUE THREAD

**Issue by schmengler** [magento/magento2#6497]
### Preconditions
1. All current Magento 2 versions (2.0, 2.1.1)
### Steps to reproduce

1. Let a commit callback throw an exception
   
   **Example:**
   - create a new search engine
   
   ```
   <type name="Magento\Search\Model\AdapterFactory">
     <arguments>
       <argument name="adapters" xsi:type="array">
         <item name="dummy" xsi:type="string">Magento\Framework\Search\Adapter\Mysql\Adapter</item>
       </argument>
     </arguments>
   </type>
   ```
   - Do not register a corresponding indexer handler for `Magento\CatalogSearch\Model\Indexer\IndexerHandlerFactory`
   - The commit callback from the fulltext indexer plugin will cause an exception "There is no such indexer handler: dummy"
2. Trigger this commit (for example, save a product)
### Expected result

1. Exception from the commit callback is thrown and can be catched later (for ex. by PHPUnit)
### Actual result

1. `rollback()` is called after `commit()` already has been called for the same transaction, resulting in an “Asymmetric transaction rollback error” exception when Magento tries to commit or rollback the last transaction on the stack.
### Possible solution
1. create a new exception type for exceptions during commit callbacks, `AfterCommitException`
2. Replace `throw $e` with `throw new AfterCommitException(null, null, $e);` in `AbstractResource::commit()`
3. In plugins that register commit callbacks, handle these exceptions separately:
   
   ```
       } catch (AfterCommitException $e) {
           throw $e->getPrevious();
       } catch (\Exception $e) {
           $productResource->rollBack();
           throw $e;
       }
   ```

Please let me know if this solution would be acceptable, before I create a PR

**Comment by mgomez-dgt** [comment]
same error help please!

[2016-11-24 19:13:22] main.INFO: Magento\Checkout\Model\PaymentInformationManagement::savePaymentInformationAndPlaceOrder [] []
[2016-11-24 19:13:22] main.INFO: ---------------------------- [] []
[2016-11-24 19:13:22] main.INFO: Asymmetric transaction rollback. [] []
[2016-11-24 19:13:22] main.INFO: Exception [] []
[2016-11-24 19:13:22] main.INFO: #0 vendor/magento/framework/Model/ResourceModel/AbstractResource.php(99): Magento\Framework\DB\Adapter\Pdo\Mysql->rollBack()
#1 var/generation/Magento/Sales/Model/ResourceModel/Order/Interceptor.php(336): Magento\Framework\Model\ResourceModel\AbstractResource->rollBack()
#2 vendor/magento/framework/Model/ResourceModel/Db/AbstractDb.php(413): Magento\Sales\Model\ResourceModel\Order\Interceptor->rollBack()
#3 vendor/magento/module-sales/Model/ResourceModel/Order.php(174): Magento\Framework\Model\ResourceModel\Db\AbstractDb->save(Object(Magento\Sales\Model\Order))
#4 vendor/magento/framework/Interception/Interceptor.php(74): Magento\Sales\Model\ResourceModel\Order->save(Object(Magento\Sales\Model\Order))
#5 vendor/magento/framework/Interception/Chain/Chain.php(70): Magento\Sales\Model\ResourceModel\Order\Interceptor->___callParent('save', Array)
#6 vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Sales\\M...', 'save', Object(Magento\Sales\Model\ResourceModel\Order\Interceptor), Array, 'clean_cache')
#7 vendor/magento/framework/App/Cache/FlushCacheByTags.php(60): Magento\Sales\Model\ResourceModel\Order\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Sales\Model\Order))
#8 vendor/magento/framework/Interception/Interceptor.php(142): Magento\Framework\App\Cache\FlushCacheByTags->aroundSave(Object(Magento\Sales\Model\ResourceModel\Order\Interceptor), Object(Closure), Object(Magento\Sales\Model\Order))
#9 var/generation/Magento/Sales/Model/ResourceModel/Order/Interceptor.php(39): Magento\Sales\Model\ResourceModel\Order\Interceptor->___callPlugins('save', Array, Array)
#10 vendor/magento/module-sales/Model/OrderRepository.php(161): Magento\Sales\Model\ResourceModel\Order\Interceptor->save(Object(Magento\Sales\Model\Order))
#11 vendor/magento/framework/Interception/Interceptor.php(146): Magento\Sales\Model\OrderRepository->save(Object(Magento\Sales\Model\Order))
#12 var/generation/Magento/Sales/Model/OrderRepository/Interceptor.php(78): Magento\Sales\Model\OrderRepository\Interceptor->___callPlugins('save', Array, Array)
#13 vendor/magento/module-sales/Model/Service/OrderService.php(191): Magento\Sales\Model\OrderRepository\Interceptor->save(Object(Magento\Sales\Model\Order))
#14 var/generation/Magento/Sales/Model/Service/OrderService/Interceptor.php(115): Magento\Sales\Model\Service\OrderService->place(Object(Magento\Sales\Model\Order))
#15 vendor/magento/module-quote/Model/QuoteManagement.php(488): Magento\Sales\Model\Service\OrderService\Interceptor->place(Object(Magento\Sales\Model\Order))
#16 vendor/magento/module-quote/Model/QuoteManagement.php(391): Magento\Quote\Model\QuoteManagement->submitQuote(Object(Magento\Quote\Model\Quote), Array)
#17 vendor/magento/module-quote/Model/QuoteManagement.php(349): Magento\Quote\Model\QuoteManagement->submit(Object(Magento\Quote\Model\Quote))
#18 vendor/magento/module-checkout/Model/PaymentInformationManagement.php(76): Magento\Quote\Model\QuoteManagement->placeOrder(319)
#19 vendor/magento/framework/Interception/Interceptor.php(146): Magento\Checkout\Model\PaymentInformationManagement->savePaymentInformationAndPlaceOrder(319, Object(Magento\Quote\Model\Quote\Payment), Object(Magento\Quote\Model\Quote\Address))
#20 var/generation/Magento/Checkout/Model/PaymentInformationManagement/Interceptor.php(26): Magento\Checkout\Model\PaymentInformationManagement\Interceptor->___callPlugins('savePaymentInfo...', Array, Array)
#21 [internal function]: Magento\Checkout\Model\PaymentInformationManagement\Interceptor->savePaymentInformationAndPlaceOrder(319, Object(Magento\Quote\Model\Quote\Payment), Object(Magento\Quote\Model\Quote\Address))
#22 vendor/magento/module-webapi/Controller/Rest.php(307): call_user_func_array(Array, Array)
#23 vendor/magento/module-webapi/Controller/Rest.php(216): Magento\Webapi\Controller\Rest->processApiRequest()
#24 var/generation/Magento/Webapi/Controller/Rest/Interceptor.php(37): Magento\Webapi\Controller\Rest->dispatch(Object(Magento\Framework\App\Request\Http))
#25 vendor/magento/framework/App/Http.php(135): Magento\Webapi\Controller\Rest\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#26 vendor/magento/framework/App/Bootstrap.php(258): Magento\Framework\App\Http->launch()
#27 index.php(44): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http))

**Comment by mgomez-dgt** [comment]
hellp please!

main.CRITICAL: Exception message: Asymmetric transaction rollback.
Trace: #0 vendor/magento/framework/Model/ResourceModel/AbstractResource.php(99): Magento\Framework\DB\Adapter\Pdo\Mysql->rollBack()
#1 var/generation/Magento/Customer/Model/ResourceModel/Customer/Interceptor.php(661): Magento\Framework\Model\ResourceModel\AbstractResource->rollBack()
#2 vendor/magento/module-customer/Model/Plugin/CustomerRepository/TransactionWrapper.php(48): Magento\Customer\Model\ResourceModel\Customer\Interceptor->rollBack()
#3 vendor/magento/framework/Interception/Interceptor.php(142): Magento\Customer\Model\Plugin\CustomerRepository\TransactionWrapper->aroundSave(Object(Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor), Object(Closure), Object(Magento\Customer\Model\Data\Customer), '********...')
#4 var/generation/Magento/Customer/Model/ResourceModel/CustomerRepository/Interceptor.php(26): Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor->___callPlugins('save', Array, Array)
#5 vendor/magento/module-customer/Model/AccountManagement.php(710): Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor->save(Object(Magento\Customer\Model\Data\Customer), '*******...')
#6 var/generation/Magento/Customer/Model/AccountManagement/Interceptor.php(141): Magento\Customer\Model\AccountManagement->createAccountWithPasswordHash(Object(Magento\Customer\Model\Data\Customer), '**********...', NULL)
#7 vendor/magento/module-customer/Model/AccountManagement.php(662): Magento\Customer\Model\AccountManagement\Interceptor->createAccountWithPasswordHash(Object(Magento\Customer\Model\Data\Customer), '*********...', NULL)
#8 var/generation/Magento/Customer/Model/AccountManagement/Interceptor.php(128): Magento\Customer\Model\AccountManagement->createAccount(Object(Magento\Customer\Model\Data\Customer), '******!', NULL)
#9 vendor/magento/module-customer/Controller/Account/CreatePost.php(277): Magento\Customer\Model\AccountManagement\Interceptor->createAccount(Object(Magento\Customer\Model\Data\Customer), '*********!', NULL)
#10 var/generation/Magento/Customer/Controller/Account/CreatePost/Interceptor.php(24): Magento\Customer\Controller\Account\CreatePost->execute()
#11 vendor/magento/framework/App/Action/Action.php(102): Magento\Customer\Controller\Account\CreatePost\Interceptor->execute()
#12 vendor/magento/framework/Interception/Interceptor.php(74): Magento\Framework\App\Action\Action->dispatch(Object(Magento\Framework\App\Request\Http))
#13 vendor/magento/framework/Interception/Chain/Chain.php(70): Magento\Customer\Controller\Account\CreatePost\Interceptor->___callParent('dispatch', Array)
#14 vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Array, 'storeCheck')
#15 vendor/magento/module-store/App/Action/Plugin/StoreCheck.php(44): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#16 vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Store\App\Action\Plugin\StoreCheck->aroundDispatch(Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#17 vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Array, 'contextPlugin')
#18 vendor/magento/module-store/App/Action/Plugin/Context.php(106): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#19 vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Store\App\Action\Plugin\Context->aroundDispatch(Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#20 vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Array, 'customer-app-ac...')
#21 vendor/magento/module-customer/Model/App/Action/ContextPlugin.php(61): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#22 vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Customer\Model\App\Action\ContextPlugin->aroundDispatch(Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#23 vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Array, 'weee-app-action...')
#24 vendor/magento/module-weee/Model/App/Action/ContextPlugin.php(112): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#25 vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Weee\Model\App\Action\ContextPlugin->aroundDispatch(Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#26 vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Array, 'tax-app-action-...')
#27 vendor/magento/module-tax/Model/App/Action/ContextPlugin.php(91): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#28 vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Tax\Model\App\Action\ContextPlugin->aroundDispatch(Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#29 vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Array, 'customer_accoun...')
#30 vendor/magento/module-customer/Controller/Plugin/Account.php(60): Magento\Customer\Controller\Account\CreatePost\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#31 vendor/magento/framework/Interception/Interceptor.php(142): Magento\Customer\Controller\Plugin\Account->aroundDispatch(Object(Magento\Customer\Controller\Account\CreatePost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#32 var/generation/Magento/Customer/Controller/Account/CreatePost/Interceptor.php(39): Magento\Customer\Controller\Account\CreatePost\Interceptor->___callPlugins('dispatch', Array, Array)
#33 vendor/magento/framework/App/FrontController.php(55): Magento\Customer\Controller\Account\CreatePost\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#34 vendor/magento/framework/Interception/Interceptor.php(74): Magento\Framework\App\FrontController->dispatch(Object(Magento\Framework\App\Request\Http))
#35 vendor/magento/framework/Interception/Chain/Chain.php(70): Magento\Framework\App\FrontController\Interceptor->___callParent('dispatch', Array)
#36 vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Framewo...', 'dispatch', Object(Magento\Framework\App\FrontController\Interceptor), Array, 'requestPreproce...')
#37 vendor/magento/module-store/App/FrontController/Plugin/RequestPreprocessor.php(94): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#38 vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#39 vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Framewo...', 'dispatch', Object(Magento\Framework\App\FrontController\Interceptor), Array, 'install')
#40 vendor/magento/framework/Module/Plugin/DbStatusValidator.php(69): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#41 vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Framework\Module\Plugin\DbStatusValidator->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#42 vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Framewo...', 'dispatch', Object(Magento\Framework\App\FrontController\Interceptor), Array, 'front-controlle...')
#43 vendor/magento/module-page-cache/Model/App/FrontController/BuiltinPlugin.php(73): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#44 vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#45 vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Framewo...', 'dispatch', Object(Magento\Framework\App\FrontController\Interceptor), Array, 'front-controlle...')
#46 vendor/magento/module-page-cache/Model/App/FrontController/VarnishPlugin.php(55): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#47 vendor/magento/framework/Interception/Interceptor.php(142): Magento\PageCache\Model\App\FrontController\VarnishPlugin->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#48 var/generation/Magento/Framework/App/FrontController/Interceptor.php(26): Magento\Framework\App\FrontController\Interceptor->___callPlugins('dispatch', Array, Array)
#49 vendor/magento/framework/App/Http.php(135): Magento\Framework\App\FrontController\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#50 vendor/magento/framework/App/Bootstrap.php(258): Magento\Framework\App\Http->launch()
#51 index.php(44): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http))
#52 {main} [] []

**Comment by tormit** [comment]
+1.
Any workaround for this issue?

**Comment by waynetheisinger** [comment]
+1. 
At the moment, none of the Integration tests work for a clean 2.1.7 build... See [this Magento Stackexchange Question](https://magento.stackexchange.com/questions/177904/magento2-why-would-magento-test-framework-integration-tests-fail-on-a-clean)

**Comment by waynetheisinger** [comment]
Hi @schmengler I just noticed that when you originally created this you asked the core team to let you know if your proposed solution was acceptable so that you could create a PR... they added an "up for grabs" label on 31st May 2017 so looks like they are happy with your solution.. if you wanted to do it? otherwise if you give me a bit more guidance I'll have a go at it.... I've got broken tests presently and its obscuring the cause...

Am I reading your proposed solution correctly - you suggest:

1. that we create a new Exception type `AfterCommitException`

2. Replace `throw $e` with `throw new AfterCommitException(null, null, $e)`; in `Magento\Framework\Model\ResourceModel\AbstractResource` 

        /**
        * Commit resource transaction
        *
        * @return $this
        * @api
        */
        public function commit()
        {
            $this->getConnection()->commit();
             /**
             * Process after commit callbacks
             */
            if ($this->getConnection()->getTransactionLevel() === 0) {
                $callbacks = CallbackPool::get(spl_object_hash($this->getConnection()));
                try {
                    foreach ($callbacks as $callback) {
                        call_user_func($callback);
                    }
                } catch (\Exception $e) {
                   throw new AfterCommitException(null, null, $e);
                }
            }
            return $this;
        }

3. In plugins that register commit callbacks, handle these exceptions separately:

        } catch (AfterCommitException $e) {
            throw $e->getPrevious();
        } catch (\Exception $e) {
            $productResource->rollBack();
            throw $e;
        }

Step three strikes me as the tricky bit - how will I locate all the commit callbacks, and make sure I don't miss any?

There are two obvious ones because they are named `addCommitCallback()`
    `\Magento\CatalogSearch\Model\Indexer\Fulltext\Plugin\Category`
   ` \Magento\CatalogSearch\Model\Indexer\Fulltext\Plugin\Product`

but how do I find other places where a rollback is called from inside catch block, and then re-throws the exception... that aren't conveniently called CommitCallback

**Comment by schmengler** [comment]
@waynetheisinger thanks for noticing! Yes, this was the idea and step three is tricky indeed. I think the best way to find the places that need to change is to look for usages of `commit()`. Commit callbacks could have been added by an extension, so for any `commit()` within a `try..catch` block  the `AfterCommitException` should be handled. Do you want to give it a try? Then please go ahead!

**Comment by waynetheisinger** [comment]
@schmengler The thinking being that we do this, even if a rollback isn't called in any of the catch blocks for the try block where the commit is called, because a rollback may be called further up the stack...?

**Comment by schmengler** [comment]
@waynetheisinger no, just that even if no callback is registered at that point like in the aforementioned plugins, there still might be callbacks. I think we can assume that a rollback will always be in a try..catch block immediately after `beginTransaction`. And exactly those are the places we are looking for.

**Comment by waynetheisinger** [comment]
@schmengler - got it - and looking though a "find" I think you're right. 

Although there are places where commit is called that are not in a `try-catch` block and there are also places where a method called `rollback` is called without a matching `commit`, these are not true rollbacks in the transaction sense (relating to things like database rollbacks in the backup system). 

So I think you are right that we can assume that a rollback will always be in a try..catch block immediately after `beginTransaction`. And that a search on commit will find them though I think I might also do a cross comparison search on `beginTransaction` too.

**Comment by waynetheisinger** [comment]
@schmengler I presume that it is only Resource Model `commits` we are interested in because they extend `AbstractResource`:

    try {
            // Find and remove unused attributes
            foreach ($attributeTables as $attributeTable) {
                $progress->setMessage($attributeTable . ' ');
                $affectedIds = $this->getAffectedAttributeIds($connection, $attributeTable);
                if (count($affectedIds) > 0) {
                    $connection->delete($attributeTable, ['value_id in (?)' => $affectedIds]);
                }
                $progress->advance();
            }
            $this->attributeResource->commit();

            $output->writeln("");
            $output->writeln("<info>Unused product attributes successfully cleaned up:</info>");
            $output->writeln("<comment>  " . implode("\n  ", $attributeTables) . "</comment>");
            return \Magento\Framework\Console\Cli::RETURN_SUCCESS;
        } catch (\Exception $exception) {
            $this->attributeResource->rollBack();

            $output->writeln("");
            $output->writeln("<error>{$exception->getMessage()}</error>");
            // we must have an exit code higher than zero to indicate something was wrong
            return \Magento\Framework\Console\Cli::RETURN_FAILURE;
        }

And we are not interested in commits acting directly on a connection:

    protected function syncData($indexer, $destinationTable, $ids)
    {
        $connection = $indexer->getConnection();
        $connection->beginTransaction();
        try {
            // remove old index
            $where = $connection->quoteInto('entity_id IN(?)', $ids);
            $connection->delete($destinationTable, $where);
            // insert new index
            $indexer->insertFromTable($indexer->getIdxTable(), $destinationTable);
            $connection->commit();
        } catch (\Exception $e) {
            $connection->rollBack();
            throw $e;
        }
    } 

Anyway, I'll proceed on this assumption, let me know if I am wrong

**Comment by waynetheisinger** [comment]
I've have created PR #9955  for this but I'm not very clear on how to move this ticket into the PR created column of the Project Board - or is this something that board moderators do? I would imagine that its a drag and drop functionality for those with permissions? (Which I mention only to stop someone else pointlessly starting work on this...)

**Comment by schmengler** [comment]
Yes. I see the team is already handling the PR. Good job!

**Comment by maghamed** [comment]
Hey, @waynetheisinger  I reviewed PR #9955 and for me proposed solution to introduce a custom exception looks too fragile.
The negative aspects I see:

- We introduce new Exception type (AfterCommitException) which MUST be thrown just in a single place in a system. This place is - https://github.com/waynetheisinger/magento2/blob/973eb084a2ae1a9dd7954e855513b76f97020c6c/lib/internal/Magento/Framework/Model/ResourceModel/AbstractResource.php#L94-L96
abstract resource through which all commit operations happen and which process callback functions. Thus, it's not allowed to throw this exception in any other place.
- If this Exception would be thrown in any other place - that could lead to unpredictable problems (both logical and programmatical)
- If we will add `catch (\Magento\Framework\Exception\AfterCommitException $e) { ... }`  to the wrong place - that could lead to unpredicatable problems, possible - assymetric transaction rollback in a wrong place or transaction would not be commied at all.

As for me all the Commit Callbacks should be independent pieces of business logic which must not affect the main execution flow. 
Like in your @schmengler  case with Full Text indexation by particular Search adapter at the time of product modification. If we are unable to proceed with re-indexation that's not a reason to roll back or affect in any way Product save process.
Ideally, Commit Collback operations are asynchronous by their nature. 

What I propose is to log critical message to the Log instead of re-throwing exception on the higher level


```php

Magento\Framework\Model\ResourceModel;

abstract class AbstractResource
{
//....
    public function commit()
    {
        $this->getConnection()->commit();
        /**
         * Process after commit callbacks
         */
        if ($this->getConnection()->getTransactionLevel() === 0) {
            $callbacks = CallbackPool::get(spl_object_hash($this->getConnection()));
            try {
                foreach ($callbacks as $callback) {
                    call_user_func($callback);
                }
            } catch (\Exception $e) {
                //throw $e;
               $this->logger->critical($e);  //Instead of throwing exception - just log exceptional case happened in After Commit Callback
            } 
        }
        return $this;
    }
}
```

If you need to notify your code about some exceptional situation happened in your After Commit Callback, you could wrap own auto commit callback with` try{}catch(){}` and handle it properly.

**Comment by waynetheisinger** [comment]
@maghamed I'd be happy with that as a solution... I did it this way because that was the proposed solution at the time the **_"Up for Grabs"_** label was added... do you want me to do the change and resubmit PR #9955?

**Comment by maghamed** [comment]
@waynetheisinger  yes, you right. I needed some time to think about this problem and a way how to solve it correctly. Because initially I also marked you PR as "approved and ready to be merged".

It would be great if you will make changes and re-submit PR.

@schmengler do you have any objections to the proposed solution?

**Comment by schmengler** [comment]
@maghamed generally I would prefer escalating exceptions as high as possible over logging them silently and move on, but in this case it's really a problem. I like your solution, it's more stable than my original proposal. 

> Ideally, Commit Collback operations are asynchronous by their nature.

that's also a good point!

**Comment by waynetheisinger** [comment]
@schmengler - I think there is a slightly confusing typo in your answer do you actually mean:

>  generally I would prefer escalating exceptions as high as possible over logging them silently and move on, but in this case it's **NOT** really a problem. I like your solution, it's more stable than my original proposal.

**Comment by schmengler** [comment]
@waynetheisinger the confusing part of my answer is, what I meant with *it*. Escalating exceptions IS a problem (the one described in the issue), so it's better to take the alternative.

**Comment by waynetheisinger** [comment]
@schmengler - ah... all becomes clear, right I understand.

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-69895

**Comment by waynetheisinger** [comment]
Hi @magento-team - you added a bunch of labels to this issue 10 hours ago but #9955 has been merged to develop and really this issue should be closed

**Comment by korostii** [comment]
That @magento-team's comment seems to be automatic. The bot is probably walking through all the internal ticket references it has, no matter the status.

Anyway, your PR being merged - doesn't actually mean it will be released anytime soon.
Things since to move rather slow around here, and it'll probably only appear in 2.2.0 (if not 2.3.0), so unfortunately it might be a bit too early to declare the issue "dealt with and forgotten" just yet =(

**Comment by dzmcclure** [comment]
I'm working in EE 2.1.7, and I've changed the line in vendor/magento/framework/Model/ResourceModel/AbstractResource.php to log the critical error instead of throwing Exception $e, however this doesn't seem to fix the issue. I've been searching for days for any explanation of what is causing the asymmetric rollback.

Is this discussion _only_ about how to properly log the error?

**Comment by waynetheisinger** [comment]
@dzmcclur the steps you've followed mean that you have eliminated plugins as the source of the rollback... you'll need to debug to find the true cause... This thread isn't the right medium for helping you. May I suggest posting on [magento.stackexchange.com](http://magento.stackexchange.com)

**Comment by ccoffee** [comment]
I have this issue, but more generically. I have installed magento 2.1.8 EE (no sample data). Saving the very first product does this. Tried on various OS'es and builds of PHP (reinstalled each time). It simply does not work.

It seems clearly tied to FullTextSearch addCommitCallback function in module-catalog-search/Model/Indexer/Fulltext/Plugin/Product.php
because preventing it makes the whole problem go away (but then your catalogsearch isn't going to update right)

**Comment by magento-engcom-team** [comment]
@schmengler, thank you for your report.
The issue is already fixed in develop branch
But we will consider to backport the fix to patch releases

**Comment by magento-engcom-team** [comment]
HI @schmengler the fix has been backported to 2.2.1 and will be available soon

**Comment by tuyennn** [comment]
I'm able to reproduce this issue on latest 2.1.14

**Comment by waynetheisinger** [comment]
@tuyennn Yes you will be able to reproduce on 2.1.4 because as @magento-engcom-team pointed out the fix has been backported only as far as 2.2.1
